mountpoint-s3/src/fs.rs

```diff
@@ -764,14 +770,22 @@ where
             FileHandleType::Write(request) => request.lock().await,
             FileHandleType::Read { .. } => return Ok(()),
         };
-        match request.complete(&file_handle.full_key).await {
+        match request.complete(&file_handle.full_key, ignore_if_empty).await {
             // According to the `fsync` man page we should return ENOSPC instead of EFBIG if it's a
             // space-related failure.
             Err(e) if e.to_errno() == libc::EFBIG => Err(err!(libc::ENOSPC, source:e, "object too big")),
             ret => ret,
         }
     }
 
+    pub async fn fsync(&self, _ino: InodeNo, fh: u64, _datasync: bool) -> Result<(), Error> {
+        self.complete_upload(fh, false).await
+    }
+
+    pub async fn flush(&self, _ino: InodeNo, fh: u64, _lock_owner: u64) -> Result<(), Error> {
+        self.complete_upload(fh, true).await
```

Comment:
probably worth a comment here summarizing why ignoring on zero length is the behavior we've chosen